garrysmod/lua/includes/extensions/vector.lua

```diff
@@ -3,10 +3,257 @@ local meta = FindMetaTable( "Vector" )
+function meta:GetZero( ... )
+	local v = Vector( self )
+	v:Zero( ... )
+	return v
+end
+
+--[[---------------------------------------------------------
+Calculates 3x3 determinant by using other two vectors as rows
+-----------------------------------------------------------]]
+function meta:Determinant( vec2, vec3 )
+	local a, b, c = self:Unpack()
+	local d, e, f = vec2:Unpack()
+	local g, h, i = vec3:Unpack()
+	return ( ( a * e * i ) + ( b * f * g ) + ( d * h * c ) - ( g * e * c ) - ( h * f * a ) - ( d * b * i ) )
+end
+
+--[[---------------------------------------------------------
+Applies new basis to a vector by the three axises of an angle
+-----------------------------------------------------------]]
+function meta:Basis( ang )
+	local x = self:Dot( ang:Forward() )
+	local y = self:Dot( ang:Right() )
+	local z = self:Dot( ang:Up() )
+	self:SetUnpacked( x, y, z )
+end
+
+--[[---------------------------------------------------------
+Applies new basis to a vector by the three axises of an angle
+-----------------------------------------------------------]]
+function meta:GetBasised( ... )
+	local v = Vector( self )
+	v:Basis( ... )
+	return v
+end
+
+--[[---------------------------------------------------------
+Projects the vector in to an axis defined by the second vector
+-----------------------------------------------------------]]
+function meta:Project( vec )
+	local vs = Vector( self )
+	self:Set( vec )
+	self:Normalize()
+	self:Mul( vs:Dot( self ) )
+end
+
+--[[---------------------------------------------------------
+Projects the vector in to an axis defined by the second vector
+-----------------------------------------------------------]]
+function meta:GetProjected( ... )
+	local v = Vector( self )
+	v:Project( ... )
+	return v
+end
+
+--[[---------------------------------------------------------
+Retrieves the angle between two vectors from 0 to math.pi
+-----------------------------------------------------------]]
+function meta:AngleBetween( vec, nrm )
+	if( nrm == nil ) then
+		return math.acos( self:Dot( vec ) / math.sqrt( self:LengthSqr() * vec:LengthSqr() ) )
+	end
+	return math.atan2( self:Determinant( vec, nrm:GetNormalized() ), self:Dot( vec ) )
+end
+
+--[[---------------------------------------------------------
+This function traces both lines and if they are not parallel
+calculates their point of intersection. Every ray is determined
+by an origin /vO/ and direction /vD/. On success returns the length
+of the closest intersect distance fraction to the interextion point.
+The first origin /vO1/ is actually the current object ( self )
+Takes:
+	vD1 --> Direction of the first ray
+	vO2 --> Position origin of the second ray
+	vD2 --> Direction of the second ray
+Returns:
+	f1 --> Intersection fraction of the first ray
+	f2 --> Intersection fraction of the second ray
+-----------------------------------------------------------]]
+function meta:IntersectFraction( vD1, vO2, vD2 )

```

Comment:
You have a bug with:
```Lua
local dn = ( dx:Length() ) ^ 2
if( dn <= 0 ) then return nil end
```
since x^2 will always be >= 0.

I don't recognise this formula - what is the "intersection fraction"? Since this takes two independent vectors and directions, I think this would be better suited for the ``math`` or ``util`` libraries.

Reply:
Should I remove it then ?

Reply:
Can you link to a mathematical/code source from where you got this formula from?

Reply:
Em, I know it from the university ... Maths part 2 analytical geometry. But I am sure that you can find it in the web also ;)

Reply:
It can be done with `== 0`, though I like <= better :D. Anywayn` is actually `|dx|^2` the length of the cross procuct of the directions. Maybe it is not needed in the vector library .Removed ;)